### PR TITLE
pagemap-cache: handle short reads

### PR DESCRIPTION
**Pull request by rst0git**
It is possible for `pread()` to return fewer number of bytes than requested. In such case, we need to repeat the read operation with appropriate offset.

Fixes: https://github.com/checkpoint-restore/criu/issues/2365

### CODE REVIEW

**Comment by avagin** [comment]
```
======================== Run zdtm/static/maps02 in uns =========================
Start test
./maps02 --pidfile=maps02.pid --outfile=maps02.out
Run criu dump
Run criu restore
=[log]=> dump/zdtm/static/maps02/117/1/restore.log
------------------------ grep Error ------------------------
b'(00.015045)      1: net: Try to restore a link 10:1:lo'
b'(00.015052)      1: net: Restoring link lo type 1'
b'(00.017672)      1: net: \tRunning ip addr restore'
b"ERROR: ld.so: object '/home/avagin/git/criu/test/ld-preload/libinterceptor.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored."
b'Error: ipv4: Address already assigned.'
b'Error: ipv6: address already assigned.'
------------------------ ERROR OVER ------------------------
Send the 15 signal to  145
Wait for zdtm/static/maps02(145) to die for 0.100000
Removing dump/zdtm/static/maps02/117
========================= Test zdtm/static/maps02 PASS =========================
```

Is this error expected?

File: test/ld-preload/libinterceptor.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#define _GNU_SOURCE
+#include <unistd.h>
+#include <dlfcn.h>
+#include <errno.h>
+
+ssize_t (*original_pread)(int fd, void *buf, size_t count, off_t offset) = NULL;
+
+ssize_t pread64(int fd, void *buf, size_t count, off_t offset)
```
Comment (avagin) [inline test/ld-preload/libinterceptor.c:8]: I think it is better to have a comment to explain what code path in criu is tested.

**Comment by avagin** [comment]
Overall, it looks good to me. Thanks.

**Comment by rst0git** [comment]
```
======================== Run zdtm/static/maps02 in uns =========================
...
b'(00.017672)      1: net: \tRunning ip addr restore'
b"ERROR: ld.so: object '/home/avagin/git/criu/test/ld-preload/libinterceptor.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored."
```
> Is this error expected?

I was able to replicate this error on Ubuntu but not in Fedora.

It looks like the reason for the error is because the repository is under `/home/avagin` and the test is running with `uns` flavor, i.e., `ip addr restore` is running in the restored user namespace and unable to access the library path. However, I believe that we can ignore this as we are testing only the code path in CRIU and there is no need to set LD_PRELOAD for the ip tool.

File: test/zdtm.py [hunk 2]
```diff
@@ -2788,6 +2805,7 @@ def get_cli_args():
                     help="Select tests for a shard <index> (0-based)")
     rp.add_argument("--test-shard-count", type=int, default=0,
                     help="Specify how many shards are being run (0=sharding disabled; must be the same for all shards)")
+    rp.add_argument("--preload", action="store_true", help="Run criu with interceptor library using LD_PRELOAD")
```
[the remark this answers is not in the document]
Comment (rst0git) [reply test/zdtm.py]: Thanks! I renamed the option to `--preload-libfault`.

**Pull request merged**